SIGSEGV in discoverPkgsFromPomProperties on parsing Java pom.properties

**What happened**:

Analysis of an existing image, that completed in grype 0.2.0 have started failing in syft code in grype 0.3.0 (syft 0.5.1)

**What you expected to happen**:

A valid analysis of the image

**How to reproduce it (as minimally and precisely as possible)**:

Unfortunately the image in question is internal and not one I can share. The invocation was from grype, and grype 0.3.0 was called with only the image name as an argument, and with -vv as an argument to produce additional debug output for this report. The error was the same for both invocations.

**Anything else we need to know?**:

The packages java application is a spring-boot fat jar containing it's own pom.properties file, as well as packages pom.properties file in the embedded dependencies. The debug output does not aid me in identifying which file it has an issue with. I am attaching an archive with the majority of the extracted dependency files related to metadata. I've stripped all code due to size, I could provide a more complete archive if necessary for those dependencies that are opensource. A number of internal artifacts have had to be removed entirely, unfortunately.

**Environment**:
- Output of `0.5.1, as called by grype 0.3.0`:

```
[0000] DEBUG Application config:
configpath: ""
presenteropt: 2
output: table
scopeopt: 1
scope: Squashed
quiet: false
log:
  structured: false
  levelopt: debug
  level: ""
  filelocation: ""
clioptions:
  configpath: ""
  verbosity: 2
db:
  dir: /root/.cache/grype/db
  updateurl: https://toolbox-data.anchore.io/grype/databases/listing.json
  autoupdate: true
dev:
  profilecpu: false
checkforappupdate: true
failon: ""
failonseverity: null

[0000] DEBUG No new grype update available
[0000]  INFO cataloging image from-lib=syft
[0000] DEBUG image: source=DockerDaemon location=<local_repository>/<project>/<project>-config:master-202006091039-26-197a51b processingOption=1 from-lib=stereoscope
[0000] DEBUG checking for available database updates
[0001] DEBUG found database update candidate: Listing(url=https://toolbox-data.anchore.io/grype/databases/vulnerability-db_v1_2020-11-06T08:21:35Z.tar.gz)
[0001] DEBUG no database update available
[0001] DEBUG image metadata: digest=sha256:e8a316f487e1e16b6827689732f4c5f5052fa899d0f75474e798a6c839d8319b mediaType=application/vnd.docker.distribution.manifest.v2+json tags=[<local_repository>/<project>/<project>-config:master-202006091039-26-197a51b] from-lib=stereoscope
[0001] DEBUG layer metadata: index=0 digest=sha256:7789f1a3d4e9258fbe5469a8d657deb6aba168d86967063e9b80ac3e1154333f mediaType=application/vnd.docker.image.rootfs.diff.tar.gzip from-lib=stereoscope
[0002] DEBUG layer metadata: index=1 digest=sha256:9e53fd4895597d04f8871a68caea4c686011e1fbd0be32e57e89ada2ea5c24c4 mediaType=application/vnd.docker.image.rootfs.diff.tar.gzip from-lib=stereoscope
[0002] DEBUG layer metadata: index=2 digest=sha256:2a19bd70fcd4ce7fd73b37b1b2c710f8065817a9db821ff839fe0b4b4560e643 mediaType=application/vnd.docker.image.rootfs.diff.tar.gzip from-lib=stereoscope
[0002] DEBUG layer metadata: index=3 digest=sha256:8891751e0a1733c5c214d17ad2b0040deccbdea0acebb963679735964d516ac2 mediaType=application/vnd.docker.image.rootfs.diff.tar.gzip from-lib=stereoscope
[0002] DEBUG layer metadata: index=4 digest=sha256:0b613548e5d2ddbad3cf45c9a67f8a1e467ea41943fc0a0651eeaf66ddd7d8cc mediaType=application/vnd.docker.image.rootfs.diff.tar.gzip from-lib=stereoscope
[0002] DEBUG layer metadata: index=5 digest=sha256:887fa2d78400b2acf8c4e8f9b98396a86025ad84bb36a720eec2fa667dafd448 mediaType=application/vnd.docker.image.rootfs.diff.tar.gzip from-lib=stereoscope
[0002] DEBUG layer metadata: index=6 digest=sha256:34a9d360086db143ef765f8588b3a18fb80b6b94447a70be5298cb75148755db mediaType=application/vnd.docker.image.rootfs.diff.tar.gzip from-lib=stereoscope
[0002] DEBUG layer metadata: index=7 digest=sha256:f923fd610f8038dd6b6757d18050e16c7308c26986977995d156e972b48eb433 mediaType=application/vnd.docker.image.rootfs.diff.tar.gzip from-lib=stereoscope
[0002] DEBUG layer metadata: index=8 digest=sha256:addc7fc11eb5414afc78c03c9b8aa91c4bf84de373939b07a5c7f72d7483326c mediaType=application/vnd.docker.image.rootfs.diff.tar.gzip from-lib=stereoscope
[0002] DEBUG layer metadata: index=9 digest=sha256:24e14eb06298353e7d75d437f0e3d32a908dbb94caabce39231b0986e258214f mediaType=application/vnd.docker.image.rootfs.diff.tar.gzip from-lib=stereoscope
[0002] DEBUG layer metadata: index=10 digest=sha256:b02110e6164691f5a3c6c4bd14df6b7ed36e276e9e160ae4ad7b7305984d8756 mediaType=application/vnd.docker.image.rootfs.diff.tar.gzip from-lib=stereoscope
[0002] DEBUG layer metadata: index=11 digest=sha256:14b508b2354623b012f351508c8ebab001b7a533878a9cdf7016b9056961272d mediaType=application/vnd.docker.image.rootfs.diff.tar.gzip from-lib=stereoscope
[0002] DEBUG layer metadata: index=12 digest=sha256:fb68de2bf9418a35d73682d1149cef14e7f30977e10e5d2fceefbc4fd3aaf454 mediaType=application/vnd.docker.image.rootfs.diff.tar.gzip from-lib=stereoscope
[0002] DEBUG layer metadata: index=13 digest=sha256:1c21c198bc2a26a64613e6b33072618d11363195dac139444bd9b5d918fd3641 mediaType=application/vnd.docker.image.rootfs.diff.tar.gzip from-lib=stereoscope
[0002] DEBUG layer metadata: index=14 digest=sha256:8bb6fa0a7405f098d766414c4bc9ae2f2ada00f7205c42e4ea6fddcf963ec69a mediaType=application/vnd.docker.image.rootfs.diff.tar.gzip from-lib=stereoscope
[0002] DEBUG layer metadata: index=15 digest=sha256:333739ac128d435f076f5bd8c68f6464640d32cb0d41d61c16244921e80a52ab mediaType=application/vnd.docker.image.rootfs.diff.tar.gzip from-lib=stereoscope
[0002] DEBUG layer metadata: index=16 digest=sha256:7c818f2b82b3eccb9c439c69fe10547ed35150ccf01b1b4c9f10fca62357bdf4 mediaType=application/vnd.docker.image.rootfs.diff.tar.gzip from-lib=stereoscope
[0002] DEBUG layer metadata: index=17 digest=sha256:5c3987ac2531c1f6f542bca2acee38937dc23c009239fd0c09440d7efade4f98 mediaType=application/vnd.docker.image.rootfs.diff.tar.gzip from-lib=stereoscope
[0002]  INFO identified distro: ubuntu 20.04 from-lib=syft
[0002]  INFO building the catalog from-lib=syft
[0002] DEBUG cataloger 'ruby-gemspec-cataloger' discovered '0' packages from-lib=syft
[0002] DEBUG cataloger 'python-package-cataloger' discovered '0' packages from-lib=syft
[0002] DEBUG cataloger 'javascript-package-cataloger' discovered '0' packages from-lib=syft
[0002] DEBUG cataloger 'dpkgdb-cataloger' discovered '115' packages from-lib=syft
[0002] DEBUG cataloger 'rpmdb-cataloger' discovered '0' packages from-lib=syft
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x10 pc=0xd945d6]

goroutine 83 [running]:
github.com/anchore/syft/syft/cataloger/java.(*archiveParser).discoverPkgsFromPomProperties(0xc0000ed730, 0x0, 0x0, 0x0, 0x11706e0, 0x1, 0xc0000ed730)
	/Users/runner/go/pkg/mod/github.com/anchore/syft@v0.5.1/syft/cataloger/java/archive_parser.go:187 +0x366
github.com/anchore/syft/syft/cataloger/java.(*archiveParser).parse(0xc0000ed730, 0x4e, 0x14b3700, 0xc000310a08, 0x1, 0xc0000ed730)
	/Users/runner/go/pkg/mod/github.com/anchore/syft@v0.5.1/syft/cataloger/java/archive_parser.go:101 +0xb4
github.com/anchore/syft/syft/cataloger/java.parseJavaArchive(0xc01cc2bbd0, 0x4e, 0x14b3700, 0xc000310a08, 0x0, 0x0, 0x0, 0x0, 0x0)
	/Users/runner/go/pkg/mod/github.com/anchore/syft@v0.5.1/syft/cataloger/java/archive_parser.go:45 +0xf1
github.com/anchore/syft/syft/cataloger/java.(*archiveParser).discoverPkgsFromNestedArchives(0xc000372540, 0xc00030b2c0, 0x0, 0x0, 0x1d15260, 0x0, 0x0)
	/Users/runner/go/pkg/mod/github.com/anchore/syft@v0.5.1/syft/cataloger/java/archive_parser.go:266 +0x543
github.com/anchore/syft/syft/cataloger/java.(*archiveParser).parse(0xc000372540, 0x28, 0x14b38a0, 0xc013a1cc60, 0x40e501, 0xc000372540)
	/Users/runner/go/pkg/mod/github.com/anchore/syft@v0.5.1/syft/cataloger/java/archive_parser.go:108 +0x147
github.com/anchore/syft/syft/cataloger/java.parseJavaArchive(0xc015715f80, 0x28, 0x14b38a0, 0xc013a1cc60, 0x0, 0x0, 0x0, 0x0, 0x0)
	/Users/runner/go/pkg/mod/github.com/anchore/syft@v0.5.1/syft/cataloger/java/archive_parser.go:45 +0xf1
github.com/anchore/syft/syft/cataloger/common.(*GenericCataloger).catalog(0xc00013eb80, 0xc01f8c93e0, 0x0, 0x0, 0x0, 0x0, 0x0)
	/Users/runner/go/pkg/mod/github.com/anchore/syft@v0.5.1/syft/cataloger/common/generic_cataloger.go:105 +0x3ec
github.com/anchore/syft/syft/cataloger/common.(*GenericCataloger).Catalog(0xc00013eb80, 0x14d4c40, 0xc0000ce178, 0xc01f9e1280, 0x2, 0x2, 0x0, 0x0)
	/Users/runner/go/pkg/mod/github.com/anchore/syft@v0.5.1/syft/cataloger/common/generic_cataloger.go:62 +0xe8
github.com/anchore/syft/syft/cataloger.Catalog(0x14d4c40, 0xc0000ce178, 0xc000365380, 0x8, 0x8, 0xc00031f180, 0x10402a0, 0xc01f74e0a0)
	/Users/runner/go/pkg/mod/github.com/anchore/syft@v0.5.1/syft/cataloger/catalog.go:46 +0x166
github.com/anchore/syft/syft.CatalogFromScope(0x1, 0x14d4c40, 0xc0000ce178, 0x109eee0, 0xc0001a8280, 0x11edd1e, 0xc, 0x2, 0xc00031f180, 0xc01f74e06b)
	/Users/runner/go/pkg/mod/github.com/anchore/syft@v0.5.1/syft/lib.go:79 +0x17b
github.com/anchore/syft/syft.Catalog(0x7ffdd803e809, 0x57, 0x1, 0x0, 0x0, 0x0, 0x0, 0x0)
	/Users/runner/go/pkg/mod/github.com/anchore/syft@v0.5.1/syft/lib.go:44 +0x30d
github.com/anchore/grype/cmd.startWorker.func1.2(0xc0000342f0, 0x7ffdd803e809, 0x57, 0xc0000ce018, 0xc0000ce020, 0xc0000ce028, 0xc0141961e0, 0xc000377980)
	/Users/runner/work/grype/grype/cmd/root.go:183 +0x82
created by github.com/anchore/grype/cmd.startWorker.func1
	/Users/runner/work/grype/grype/cmd/root.go:181 +0x227
```

- OS (e.g: `cat /etc/os-release` or similar):
```
PRETTY_NAME="Debian GNU/Linux 10 (buster)"
NAME="Debian GNU/Linux"
VERSION_ID="10"
VERSION="10 (buster)"
VERSION_CODENAME=buster
ID=debian
HOME_URL="https://www.debian.org/"
SUPPORT_URL="https://www.debian.org/support"
BUG_REPORT_URL="https://bugs.debian.org/"
```
[failing-on-parsing-pom-properties.tar.gz](https://github.com/anchore/syft/files/5501421/failing-on-parsing-pom-properties.tar.gz)

--- a/syft/cataloger/java/archive_parser.go
+++ b/syft/cataloger/java/archive_parser.go
@@ -184,7 +184,7 @@ func (j *archiveParser) discoverPkgsFromPomProperties(parentPkg *pkg.Package) ([
 
 				// keep the artifact name within the virtual path if this package does not match the parent package
 				vPathSuffix := ""
-				if !strings.HasPrefix(propsObj.ArtifactID, parentPkg.Name) {
+				if parentPkg != nil && !strings.HasPrefix(propsObj.ArtifactID, parentPkg.Name) {
 					vPathSuffix += ":" + propsObj.ArtifactID
 				}
 				virtualPath := j.virtualPath + vPathSuffix
@@ -208,30 +208,34 @@ func (j *archiveParser) discoverPkgsFromPomProperties(parentPkg *pkg.Package) ([
 				// the name/version pair matches...
 				matchesParentPkg := pkgKey == parentKey
 
-				// the virtual path matches...
-				matchesParentPkg = matchesParentPkg || parentPkg.Metadata.(pkg.JavaMetadata).VirtualPath == virtualPath
+				if parentPkg != nil {
+					// the virtual path matches...
+					matchesParentPkg = matchesParentPkg || parentPkg.Metadata.(pkg.JavaMetadata).VirtualPath == virtualPath
 
-				// the pom artifactId has the parent name or vice versa
-				if propsObj.ArtifactID != "" {
-					matchesParentPkg = matchesParentPkg || strings.Contains(parentPkg.Name, propsObj.ArtifactID) || strings.Contains(propsObj.ArtifactID, parentPkg.Name)
+					// the pom artifactId has the parent name or vice versa
+					if propsObj.ArtifactID != "" {
+						matchesParentPkg = matchesParentPkg || strings.Contains(parentPkg.Name, propsObj.ArtifactID) || strings.Contains(propsObj.ArtifactID, parentPkg.Name)
+					}
+
+					if matchesParentPkg {
+						// we've run across more information about our parent package, add this info to the parent package metadata
+						// the pom properties is typically a better source of information for name and version than the manifest
+						if p.Name != parentPkg.Name {
+							parentPkg.Name = p.Name
+						}
+						if p.Version != parentPkg.Version {
+							parentPkg.Version = p.Version
+						}
+
+						parentMetadata, ok := parentPkg.Metadata.(pkg.JavaMetadata)
+						if ok {
+							parentMetadata.PomProperties = propsObj
+							parentPkg.Metadata = parentMetadata
+						}
+					}
 				}
 
-				if matchesParentPkg {
-					// we've run across more information about our parent package, add this info to the parent package metadata
-					// the pom properties is typically a better source of information for name and version than the manifest
-					if p.Name != parentPkg.Name {
-						parentPkg.Name = p.Name
-					}
-					if p.Version != parentPkg.Version {
-						parentPkg.Version = p.Version
-					}
-
-					parentMetadata, ok := parentPkg.Metadata.(pkg.JavaMetadata)
-					if ok {
-						parentMetadata.PomProperties = propsObj
-						parentPkg.Metadata = parentMetadata
-					}
-				} else if !j.discoveredPkgs.Contains(pkgKey) {
+				if !matchesParentPkg && !j.discoveredPkgs.Contains(pkgKey) {
 					// only keep packages we haven't seen yet (and are not related to the parent package)
 					pkgs = append(pkgs, p)
 				}

--- a/syft/cataloger/java/archive_parser.go
+++ b/syft/cataloger/java/archive_parser.go
@@ -178,67 +178,69 @@ func (j *archiveParser) discoverPkgsFromPomProperties(parentPkg *pkg.Package) ([
 			return nil, fmt.Errorf("failed to parse pom.properties (%s): %w", j.virtualPath, err)
 		}
 
-		if propsObj != nil {
-			if propsObj.Version != "" && propsObj.ArtifactID != "" {
-				// TODO: if there is no parentPkg (no java manifest) one of these poms could be the parent. We should discover the right parent and attach the correct info accordingly to each discovered package
+		if propsObj == nil {
+			continue
+		}
 
-				// keep the artifact name within the virtual path if this package does not match the parent package
-				vPathSuffix := ""
-				if parentPkg != nil && !strings.HasPrefix(propsObj.ArtifactID, parentPkg.Name) {
-					vPathSuffix += ":" + propsObj.ArtifactID
+		if propsObj.Version != "" && propsObj.ArtifactID != "" {
+			// TODO: if there is no parentPkg (no java manifest) one of these poms could be the parent. We should discover the right parent and attach the correct info accordingly to each discovered package
+
+			// keep the artifact name within the virtual path if this package does not match the parent package
+			vPathSuffix := ""
+			if parentPkg != nil && !strings.HasPrefix(propsObj.ArtifactID, parentPkg.Name) {
+				vPathSuffix += ":" + propsObj.ArtifactID
+			}
+			virtualPath := j.virtualPath + vPathSuffix
+
+			// discovered props = new package
+			p := pkg.Package{
+				Name:         propsObj.ArtifactID,
+				Version:      propsObj.Version,
+				Language:     pkg.Java,
+				Type:         pkg.JavaPkg,
+				MetadataType: pkg.JavaMetadataType,
+				Metadata: pkg.JavaMetadata{
+					VirtualPath:   virtualPath,
+					PomProperties: propsObj,
+					Parent:        parentPkg,
+				},
+			}
+
+			pkgKey := uniquePkgKey(&p)
+
+			// the name/version pair matches...
+			matchesParentPkg := pkgKey == parentKey
+
+			if parentPkg != nil {
+				// the virtual path matches...
+				matchesParentPkg = matchesParentPkg || parentPkg.Metadata.(pkg.JavaMetadata).VirtualPath == virtualPath
+
+				// the pom artifactId has the parent name or vice versa
+				if propsObj.ArtifactID != "" {
+					matchesParentPkg = matchesParentPkg || strings.Contains(parentPkg.Name, propsObj.ArtifactID) || strings.Contains(propsObj.ArtifactID, parentPkg.Name)
 				}
-				virtualPath := j.virtualPath + vPathSuffix
 
-				// discovered props = new package
-				p := pkg.Package{
-					Name:         propsObj.ArtifactID,
-					Version:      propsObj.Version,
-					Language:     pkg.Java,
-					Type:         pkg.JavaPkg,
-					MetadataType: pkg.JavaMetadataType,
-					Metadata: pkg.JavaMetadata{
-						VirtualPath:   virtualPath,
-						PomProperties: propsObj,
-						Parent:        parentPkg,
-					},
-				}
-
-				pkgKey := uniquePkgKey(&p)
-
-				// the name/version pair matches...
-				matchesParentPkg := pkgKey == parentKey
-
-				if parentPkg != nil {
-					// the virtual path matches...
-					matchesParentPkg = matchesParentPkg || parentPkg.Metadata.(pkg.JavaMetadata).VirtualPath == virtualPath
-
-					// the pom artifactId has the parent name or vice versa
-					if propsObj.ArtifactID != "" {
-						matchesParentPkg = matchesParentPkg || strings.Contains(parentPkg.Name, propsObj.ArtifactID) || strings.Contains(propsObj.ArtifactID, parentPkg.Name)
+				if matchesParentPkg {
+					// we've run across more information about our parent package, add this info to the parent package metadata
+					// the pom properties is typically a better source of information for name and version than the manifest
+					if p.Name != parentPkg.Name {
+						parentPkg.Name = p.Name
+					}
+					if p.Version != parentPkg.Version {
+						parentPkg.Version = p.Version
 					}
 
-					if matchesParentPkg {
-						// we've run across more information about our parent package, add this info to the parent package metadata
-						// the pom properties is typically a better source of information for name and version than the manifest
-						if p.Name != parentPkg.Name {
-							parentPkg.Name = p.Name
-						}
-						if p.Version != parentPkg.Version {
-							parentPkg.Version = p.Version
-						}
-
-						parentMetadata, ok := parentPkg.Metadata.(pkg.JavaMetadata)
-						if ok {
-							parentMetadata.PomProperties = propsObj
-							parentPkg.Metadata = parentMetadata
-						}
+					parentMetadata, ok := parentPkg.Metadata.(pkg.JavaMetadata)
+					if ok {
+						parentMetadata.PomProperties = propsObj
+						parentPkg.Metadata = parentMetadata
 					}
 				}
+			}
 
-				if !matchesParentPkg && !j.discoveredPkgs.Contains(pkgKey) {
-					// only keep packages we haven't seen yet (and are not related to the parent package)
-					pkgs = append(pkgs, p)
-				}
+			if !matchesParentPkg && !j.discoveredPkgs.Contains(pkgKey) {
+				// only keep packages we haven't seen yet (and are not related to the parent package)
+				pkgs = append(pkgs, p)
 			}
 		}
 	}

--- a/test/integration/java_no_main_package_test.go
+++ b/test/integration/java_no_main_package_test.go
@@ -1,0 +1,20 @@
+package integration
+
+import (
+	"github.com/anchore/stereoscope/pkg/imagetest"
+	"github.com/anchore/syft/syft"
+	"github.com/anchore/syft/syft/source"
+	"testing"
+)
+
+func TestJavaNoMainPackage(t *testing.T) { // Regression: https://github.com/anchore/syft/issues/252
+	fixtureImageName := "image-java-no-main-package"
+	_, cleanup := imagetest.GetFixtureImage(t, "docker-archive", fixtureImageName)
+	tarPath := imagetest.GetFixtureImageTarPath(t, fixtureImageName)
+	defer cleanup()
+
+	_, _, _, err := syft.Catalog("docker-archive:"+tarPath, source.SquashedScope)
+	if err != nil {
+		t.Fatalf("failed to catalog image: %+v", err)
+	}
+}

--- a/test/integration/test-fixtures/image-java-no-main-package/Dockerfile
+++ b/test/integration/test-fixtures/image-java-no-main-package/Dockerfile
@@ -1,0 +1,23 @@
+FROM jenkins:2.60.3
+
+USER root
+
+WORKDIR /usr/share/jenkins
+
+RUN mkdir tmp
+
+WORKDIR /usr/share/jenkins/tmp
+
+RUN apt-get update 2>&1 > /dev/null && apt-get install -y less zip 2>&1 > /dev/null
+
+RUN unzip ../jenkins.war 2>&1 > /dev/null
+
+RUN rm -f ./META-INF/MANIFEST.MF
+
+WORKDIR /usr/share/jenkins
+
+RUN rm -rf jenkins.war
+
+RUN cd ./tmp && zip -r ../jenkins.war . && cd ..
+
+RUN rm -rf ./tmp

--- a/test/integration/test-fixtures/image-large-apk-data/Dockerfile
+++ b/test/integration/test-fixtures/image-large-apk-data/Dockerfile
@@ -1,5 +1,5 @@
 FROM alpine@sha256:d9a7354e3845ea8466bb00b22224d9116b183e594527fb5b6c3d30bc01a20378
 RUN apk add --no-cache \
-            tzdata=2020f-r0 \
+            tzdata=2021a-r0 \
             vim=8.2.2320-r0 \
             alpine-sdk=1.0-r0

